### DISCUSSION TITLE
How can I create an "inner bevel" effect on text?

### DISCUSSION THREAD

**Discussion by rambouillet** [SixLabors/ImageSharp#1579] [Q&A] [answered]
I'm trying to create the following effect with some text:

![image](https://user-images.githubusercontent.com/20598632/111856882-00614980-88eb-11eb-9b7b-163156369d54.png)

Here's a method I wrote to create the text so far:

    using SixLabors.ImageSharp;
    using SixLabors.ImageSharp.Drawing.Processing;
    using SixLabors.ImageSharp.Formats.Png;
    using SixLabors.ImageSharp.Processing;
    using SixLabors.ImageSharp.PixelFormats;

    public static void CreateTextPng(string SavePath, int Width, string Text)
    {
        var TheFont = new Font(SystemFonts.Find("Times New Roman"), 40, FontStyle.Bold);
        var Size = TextMeasurer.Measure(Text, new RendererOptions(TheFont) { WrappingWidth = Width });
        using (var TheImage = new Image<Rgba32>(Width, (int)Math.Round(Size.Height)))
        {
            TheImage.Mutate(delegate (IImageProcessingContext Processor)
            {
                var textGraphicsOptions = new TextGraphicsOptions(new GraphicsOptions(), new TextOptions
                {
                    HorizontalAlignment = HorizontalAlignment.Center,
                    VerticalAlignment = VerticalAlignment.Center,
                    WrapTextWidth = TheImage.Width
                });
                Processor.DrawText(textGraphicsOptions, Text, TheFont, Color.Yellow, new PointF(0, TheImage.Height / 2));
                //
                //I think this is where the code to create the "inner bevel" effect would go, but I don't know what to put here...
                //
            });
            TheImage.SaveAsPng(SavePath, new PngEncoder { CompressionLevel = PngCompressionLevel.BestCompression });
        }
    }

And here is the method call:

    CreateTextPng("C:/dev/textexample.png", 400, "Good Omens: The Nice and Accurate Prophecies of Agnes Nutter, Witch");

Can this bevel effect be done with ImageSharp at all? Any help would be appreciated!

**Comment by JimBobSquarePants** [comment, marked as answer]
@rambouillet There's no built in methods to do this I'm afraid. 

I've done a little bit of research though into how this would be achieved. 

Bevel effects can be achieved by the application of [Bump](https://en.wikipedia.org/wiki/Bump_mapping) or [Normal](https://en.wikipedia.org/wiki/Normal_mapping) Maps to a 2D image. You'd need an [emboss convolution filter](https://en.wikipedia.org/wiki/Image_embossing) that can be generated and [rotated](https://stackoverflow.com/questions/40359452/rotating-matrix-by-n-degrees) by an arbitrary angle that is used to populate the map. 

This library which is based on an old version of ImageSharp contains [code](https://github.com/JaCraig/Structure.Sketching/tree/3ef7cf83a825b7bd962dfe6682475e8fceade597/Structure.Sketching/Filters/Pipelines) to generate maps. It shouldn't be too hard to port that using an improved emboss filter.

I'm not 100% sure how the map is then applied to the image though to alter the brightness but this link seems to contain further information.
http://www.inf.ed.ac.uk/teaching/courses/cg/lectures/cg12_2012.pdf

P.S. Love that book!
